pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -207,16 +207,21 @@ in package-set { inherit pkgs lib callPackage; } self // {
          });
 
     # Creates a Haskell package from a source package by calling cabal2nix on the source.
-    callCabal2nixWithOptions = name: src: extraCabal2nixOptions: args:
+    callCabal2nixWithOptions = name: src: opts: args:
       let
-        filter = path: type:
+        checkOpts = x@{extraCabal2nixOptions ? "", srcModifier ? null}: { inherit extraCabal2nixOptions srcModifier; };
+        checkedOpts = if builtins.isString opts then checkOpts { extraCabal2nixOptions = opts; } else checkOpts opts;
+        extraCabal2nixOptions = checkedOpts.extraCabal2nixOptions;
+        srcModifier = checkedOpts.srcModifier;
```

Comment:
I feel like this whole section could be simplified by e.g. using `or` or `inherit` syntax.